lib/pbench/test/unit/server/test_datasets_metadata.py

```diff
@@ -410,36 +404,43 @@ def test_put_invalid_deletion(self, query_get_as, query_put_as):
         }
 
     def test_put_set_errors(self, monkeypatch, query_get_as, query_put_as):
-        real_set = Metadata.setvalue
-
-        @staticmethod
-        def fake_set(dataset: Dataset, key: str, value: Any, user_id: str):
-            if key in ("global.dashboard.seen", "dataset.name"):
-                raise MetadataBadStructure(dataset, key, key)
-            real_set(dataset, key, value, user_id)
-
-        monkeypatch.setattr(Metadata, "setvalue", fake_set)
+        """Test a partial success. We set a scalar value on a key and then try
+        to set a nested value: i.e., with "global.dashboard.nested = False", we
+        attempt to set "global.dashboard.nested.dummy". We expect this to fail,
+        but we expect other values to succeed. This partial success is reported
+        as an "internal error" (for lack of any better representation in HTTP)
+        with a message indicating the specific failures.
+        """
+        query_put_as(
+            "drb",
+            {"metadata": {"global.dashboard.nested": False}},
+            "drb",
+            HTTPStatus.OK,
+        )
         response = query_put_as(
             "drb",
             {
                 "metadata": {
                     "global.dashboard.seen": False,
+                    "global.dashboard.nested.dummy": True,
                     "dataset.name": "foo",
-                    "global.dashboard.saved": True,
                 }
             },
             "drb",
             HTTPStatus.INTERNAL_SERVER_ERROR,
         )
         assert response.json == {
-            "message": "Key 'dataset.name' value for 'dataset.name' in (3)|drb is not a JSON object, Key 'global.dashboard.seen' value for 'global.dashboard.seen' in (3)|drb is not a JSON object"
+            "message": "Key 'nested' value for 'global.dashboard.nested.dummy' in (3)|foo is not a JSON object"
         }
```

Comment:
This doesn't demonstrate or exercise the "list feature", now -- it's reporting only one error.  It seems like it should be rigged to report at least two errors.

Reply:
I thought about that and decided it wasn't worth the effort. The repetition is essentially just testing that `", ".join` works, which is outside our control. We've already proven that the error is appended to the list, and that we generate a string from that list.

*Bizarre, by the way: GitHub doesn't show this comment in the `Files changed` view, no matter how many times I refresh, and if I select it in the `Conversations` menu GitHub just ignores me. I can only get to it from here. Sometimes replying to a comment on the Conversations tab gets joined with an open review and sometimes it gets posted solo: I guess we'll see what it chooses to do today when I click `Comment`...*

Reply:
> GitHub doesn't show this comment in the Files changed view, no matter how many times I refresh

That's because the conversation is attached to a diff-pair which is different from the one that you are refreshing.

I'm replying to this in the "Conversation" tab-page in the main PR page; when I click on the link above the cited diff which contains the filename ([lib/pbench/test/unit/server/test_datasets_metadata.py](https://github.com/distributed-system-analysis/pbench/pull/3091/files/123a5c237bf3ea079778dd87cd9981c713160f8b..f7d0287e6930510792143edda98d11ac437d4b79#diff-5eab7892d84f8abd63da546c67b9f0ad505dff15fc27eaeaaeb99919b298231f)) it takes me to the "Files changed" tab-page; if I click on the "Changes from" dropdown (which is labeled _Changes from 1 commit_), it shows that only one commit is selected.  But, if I scroll down a bit, I find this conversation at the cited line numbers.

In the past, I believe that conversations attached to the most recent commit (or to a file which hasn't been changed since the conversation started) have shown up in the in the "Files changed" tab when selecting all commits (which is the default), but, in this case, it's not showing up for me, either.  Curious.

I find that the "Conversations" dropdown is only useful as a reference for what is unresolved (and, by mousing over the entries, I can see _when_ they were started).  To reliably _find_ the conversations, I go to the "Conversation" tab-page and scroll to the appropriate point in the timeline.

Reply:
> Sometimes replying to a comment on the Conversations tab gets joined with an open review and sometimes it gets posted solo: I guess we'll see what it chooses to do today when I click Comment...

My experience with that is that a review is a highly stateful (on the GitHub server side) operation.  Once you start a review, then, best case, all comments you enter (in either tab) become part of the review.  (In the worst case, entering a comment in the Conversation tab causes the review to be posted, whether you were ready for that or not....)

So, when I'm doing a review, and I'm using multiple windows (which I typically do), I refresh the windows every time I change the state, so that my client accurately reflects the server's state.  (And, I try not to enter any new comments on the Conversation tab until a post the review in the Files changed tab.)

Reply:
Back to the main discussion...
>I thought about that and decided it wasn't worth the effort. The repetition is essentially just testing that ", ".join works, which is outside our control. We've already proven that the error is appended to the list, and that we generate a string from that list.

I think the repetition does two things.  Yes, it proves that `join()` works, but, more importantly, it proves that what join is working from is a list with multiple entries.  To wit, it shows that the second entry was appended and didn't overwrite the first, and that our code is working as desired.

But, it's not that big of a deal, so I shall resolve this conversation.  🙂